lib/mix/lib/mix.ex

```diff
@@ -1033,13 +1039,34 @@ defmodule Mix do
     File.rm_rf(dep_path)
   end
 
-  defp install_project_dir(cache_id) do
+  defp install_project_dir(cache_id), do: Path.join([version_dir(), cache_id])
+
+  defp latest_cache_dir(id_without_deps) do
+    version_dir = version_dir()
+
+    with {:ok, entries} <- File.ls(version_dir) do
+      entries
+      |> Enum.filter(&String.starts_with?(&1, id_without_deps))
+      |> Enum.map(&Path.join(version_dir, &1))
+      |> Enum.map(&{&1, File.stat!(&1).mtime})
+      |> Enum.max_by(&elem(&1, 1), fn -> nil end)
+      |> case do
+        {latest_dir, _mtime} -> latest_dir
+        nil -> nil
+      end
```

Comment:
Maybe something like this would be easier to read?
I lost track on 1052 I'm just guessing what should be the result.
```suggestion
      for entry <- entries,
          String.starts_with?(entry, id_without_deps),
          current_path = Path.join(version_dir, entry),
          %{mtime: current_mtime} = File.stat!(current_path),
          reduce: {nil, nil} do
        {nil, nil} -> {current_path, current_mtime}
        {_path, mtime} when current_mtime > mtime -> {current_path, current_mtime}
        acc -> acc
      end
      |> elem(0)
```

Reply:
TBH for me, I like pipe operator more, but thank you for advice.